src/Payments.sol

```diff
@@ -0,0 +1,572 @@
+                railId,
+                rail,
+                rateQueue,
+                currentRate,
+                currentSettlementEpoch,
+                settlementTargetEpoch
+            );
+        }
+
+        // Update account balances.
+        payer.funds -= totalSettledAmount;
+        payee.funds += totalSettledAmount;
+        require(payer.lockupCurrent >= totalSettledAmount, "Insufficient lockup funds");
+        payer.lockupCurrent -= totalSettledAmount;
+
+        // Record the new settlement epoch on the rail.
+        rail.lastSettledAt = finalSettledEpoch;
+
+        require(payer.lockupCurrent <= payer.funds, "insufficient funds");
+
+        return (totalSettledAmount, finalSettledEpoch);
+    }
+
+    function _settleSegment(
+        uint256 railId,
+        Rail storage rail,
+        uint256 segmentStart,
+        uint256 segmentEnd,
+        uint256 rate
+    ) internal returns (uint256 settledAmount, uint256 settledUntil) {
+        // Calculate the intended payment for the entire segment.
+        uint256 duration = segmentEnd - segmentStart;
+        settledAmount = rate * duration;
+        settledUntil = segmentEnd;
+
+        if (rail.arbiter != address(0)) {
+            // Call the external arbitrator.
+            IArbiter arbiter = IArbiter(rail.arbiter);
+            IArbiter.ArbitrationResult memory arbResult = arbiter.arbitratePayment(
+                railId,
+                settledAmount,
+                segmentStart,
+                segmentEnd
+            );
+            require(arbResult.approved, "arbitrer refused payment");
+            // Do not settle past the segment end.
+            settledUntil = arbResult.settledUntil;
+            settledAmount = arbResult.modifiedAmount;
+        }
+
+        // If no progress is made, revert.
+        require(settledUntil > segmentStart, "no progress made in settlement");
+
+        return (settledAmount, settledUntil);
+    }
+    function _settleWithRateChanges(
+        uint256 railId,
+        Rail storage rail,
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 currentRate,
+        uint256 startEpoch,
+        uint256 targetEpoch
+    ) internal returns (uint256 totalSettled, uint256 finalEpoch) {
+        totalSettled = 0;
+        uint256 currentEpoch = startEpoch;
+
+        while (currentEpoch < targetEpoch) {
+            // Determine the next boundary: either the next queued rate–change or the targetEpoch.
+            uint256 nextBoundary = targetEpoch;
+            uint256 segmentRate = currentRate;
+            if (!rateQueue.isEmpty()) {
+                RateChangeQueue.RateChange memory nextChange = rateQueue.peek();
+                if (nextChange.untilEpoch > currentEpoch && nextChange.untilEpoch <= targetEpoch) {
+                    nextBoundary = nextChange.untilEpoch;
+                    segmentRate = nextChange.rate;
+                }
```

Comment:
I've spent a lot of time staring at this code and I'm still not sure if it's right. Let's talk through it tomorrow.

Reply:
I agree. It is involved.

Reply:
Let me try to simplify it more today before we sync up.

Reply:
Have refactored this entire flow to make it easy to reason about.